File: 2016/Day_2/day2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func solutionPart1(instructions string, button int) int {
	for _, instruct := range instructions {
		switch string(instruct) {
		case "U":
			if button > 3 {
				button -= 3
			}
		case "D":
			if button < 7 {
				button += 3
			}
		case "L":
			if button != 1 && button != 4 && button != 7 {
				button -= 1
			}
		case "R":
			if button != 3 && button != 6 && button != 9 {
				button += 1
			}
		}
	}
	return button
}

func moveUp(button string) string {
	up := []string{"1", "2", "4", "5", "9"} // can't go up from these buttons
	for _, val := range up {
		if val == button {
			return button
		}
	}
	if button == "A" {
		return "6"
	} else if button == "B" {
		return "7"
	} else if button == "C" {
		return "8"
	} else if button == "D" {
		return "B"
	} else {
		i_button, err := strconv.Atoi(button)
		if err != nil {
			fmt.Println(err)
		}
		if i_button-4 < 0 {
			return strconv.Itoa(-(i_button - 4))
		} else {
			return strconv.Itoa(i_button - 4)
		}
	}
}

func moveDown(button string) string {
	down := []string{"5", "9", "A", "C", "D"}
	for _, val := range down {
		if val == button {
			return button
		}
	}
	if button == "6" {
		return "A"
	} else if button == "7" {
		return "B"
	} else if button == "8" {
		return "C"
	} else if button == "B" {
		return "D"
	} else if button == "1" {
		return "3"
	} else {
		i_button, err := strconv.Atoi(button)
		if err != nil {
			fmt.Println(err)
		}
		return strconv.Itoa(i_button + 4)
	}
}

func moveLeft(button string) string {
	left := []string{"1", "2", "5", "A", "D"}
	for _, val := range left {
		if val == button {
			return button
		}
	}
	if button == "B" {
		return "A"
	} else if button == "C" {
		return "B"
	} else {
		i_button, err := strconv.Atoi(button)
		if err != nil {
			fmt.Println(err)
		}
		return strconv.Itoa(i_button - 1)
	}
}

func moveRight(button string) string {
	right := []string{"1", "4", "9", "C", "D"}
	for _, val := range right {
		if val == button {
			return button
		}
	}
	if button == "A" {
		return "B"
	} else if button == "B" {
		return "C"
	} else {
		i_button, err := strconv.Atoi(button)
		if err != nil {
			fmt.Println(err)
		}
		return strconv.Itoa(i_button + 1)
	}
}

func solutionPart2(instructions string, button string) string {
	for _, instruct := range instructions {
		switch string(instruct) {
		case "U":
			button = moveUp(button)
		case "D":
			button = moveDown(button)
		case "L":
			button = moveLeft(button)
		case "R":
			button = moveRight(button)
		}
	}
	return button
}

func main() {
	path := "puzzle_input.txt"
	// test := base + "test_input.txt"
	file, err := ioutil.ReadFile(path)
	if err != nil {
		fmt.Println(err)
	}
	instructions := strings.Split(string(file), "\n")

	button1 := 5
	password1 := make([]int, 0)
	button2 := "5"
	password2 := make([]string, 0)
	for _, instruct := range instructions {
		button1 = solutionPart1(instruct, button1)
		password1 = append(password1, button1)
		button2 = solutionPart2(instruct, button2)
		password2 = append(password2, button2)
	}

	fmt.Println("Part 1:", password1)
	fmt.Println("Part 2:", password2)
}

File: 2016/Day_2/day2.java
import java.io.File;
import java.util.Scanner;

public class day2 {

    public static void main(String[] args) 
    {
        String path = "puzzle_input.txt";
        File myFile = new File(path);
        String[] instructions = new String[5];
        int index = 0;
        try {
            Scanner scanner = new Scanner(myFile);
            while (scanner.hasNextLine()) {
                instructions[index++] = scanner.nextLine();
            }
            scanner.close();
        } catch (Exception e) {
            System.out.println("ERROR:");
            e.printStackTrace();
            System.exit(1);
        }

        int button1 = 5;
        String password1 = new String();
        char button2 = '5';
        String password2 = new String();
        for (int i = 0; i < instructions.length; i++) {
            button1 = solutionPart1(instructions[i], button1);
            password1 +=  String.valueOf(button1);
            button2 = solutionPart2(instructions[i], button2);
            password2 += String.valueOf(button2);
        }
        System.out.println("Part 1: " + password1);
        System.out.println("Part 2: " + password2);
    }
    
    static int solutionPart1(String instructions, int button) 
    {
        for (int i = 0 ; i < instructions.length(); i++) {
            char current = instructions.charAt(i);
            if (current == 'U') {
                if (button > 3) {
                    button -= 3;
                }
            } else if (current == 'D') {
                if (button < 7) {
                    button += 3;
                }
            } else if (current == 'L') {
                if (button != 1 && button != 4 && button != 7) {
                    button--;
                }
            } else {    // 'R'
                if (button != 3 && button != 6 && button != 9) {
                    button++;
                }
            }
        }
        return button;
    }

    static int[] findButtonIndex(char[][] keypad, char button) 
    {
        for (int i = 0; i < keypad.length; i++) {
            for (int j = 0; j < keypad[i].length; j++) {
                if (keypad[i][j] == button) {
                    return new int[]{i, j};
                }
            }
        }
        return null;
    }

    static char solutionPart2(String instructions, char button) 
    {
        char[][] keypad = new char[][] {
            {' ', ' ', '1', ' ', ' '},
            {' ', '2', '3', '4', ' '},
            {'5', '6', '7', '8', '9'},
            {' ', 'A', 'B', 'C', ' '},
            {' ', ' ', 'D', ' ', ' '}  };
        int[] index = findButtonIndex(keypad, button);
        if (index == null) {
            System.out.println("ERROR: null pointer");
            System.exit(1);
        }
        int row = index[0], col = index[1];
        for (int i = 0; i < instructions.length(); i++) {
            char current = instructions.charAt(i);
            if (current == 'U') {
                if (row - 1 >= 0 && keypad[row - 1][col] != ' ') {
                    row--;
                }
            } else if (current == 'D') {
                if (row + 1 < keypad.length && keypad[row + 1][col] != ' ') {
                    row++;
                }
            } else if (current == 'R') {
                if (col + 1 < keypad[row].length && keypad[row][col + 1] != ' ') {
                    col++;
                }
            } else {    // 'L' 
                if (col - 1 >= 0 && keypad[row][col - 1] != ' ') {
                    col--;
                }
            }
        }
        return keypad[row][col];
    }
}
